optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```
@@ -867,6 +867,146 @@
+     * </li>
+     * <li>Simplicity.
+     * The function should only take the argument and return another simple value,
+     * without performing any significant computation.
+     * Not following this recommendation may result in performance loss,
+     * as the function may be called frequently.</li>
+     * <li>Bijectivity.
+     * No two input tuples should map to the same output tuple,
+     * or to tuples that are {@link Object#equals(Object) equal}.
+     * Not following this recommendation will create a constraint stream with duplicate tuples,
+     * and may force you to use {@link #distinct()} later.</li>
+     * <li>Immutable data carriers.
+     * The tuples returned by the mapping function should be identified by their contents and nothing else.
+     * They should also be immutable, meaning their contents should not be allowed to change.
+     * If two tuples carry objects which {@link Object#equals(Object) equal} one another,
+     * those two tuples should likewise {@link Object#equals(Object) equal} and preferably be the same instance.
+     * Java Records are the ideal tool for this particular use case.</li>
+     * </ul>
+     *
+     * <p>
+     * Simple example: assuming a constraint stream of tuple of {@code Person}s
+     * {@code [Ann(age = 20), Beth(age = 25), Cathy(age = 30)]},
+     * calling {@code map(Person::getAge)} on such stream will produce a stream of {@link Integer}s
+     * {@code [20, 25, 30]},
+     *
+     * <p>
+     * Example with a non-bijective mapping function: assuming a constraint stream of tuple of {@code Person}s
+     * {@code [Ann(age = 20), Beth(age = 25), Cathy(age = 30), David(age = 30), Eric(age = 20)]},
+     * calling {@code map(Person::getAge)} on such stream will produce a stream of {@link Integer}s
+     * {@code [20, 25, 30, 30, 20]}.
+     *
+     * @param mapping never null, function to convert the original tuple into the new tuple
```

I haven't noticed any null check to fail fast if this precondition is not met in the implementation code.